serde-generate/src/ocaml.rs

```diff
@@ -0,0 +1,354 @@
+use crate::{
+    indent::{IndentConfig, IndentedWriter},
+    CodeGeneratorConfig,
+};
+use heck::CamelCase;
+use heck::SnakeCase;
+use include_dir::include_dir as include_directory;
+use phf::phf_set;
+use serde_reflection::{ContainerFormat, Format, Named, Registry, VariantFormat};
+use std::{
+    collections::BTreeMap,
+    io::{Result, Write},
+    path::PathBuf,
+};
+
+fn capitalize(s: &str) -> String {
+    let mut c = s.chars();
+    match c.next() {
+        None => String::new(),
+        Some(f) => f.to_uppercase().collect::<String>() + c.as_str(),
+    }
+}
+
+pub struct CodeGenerator<'a> {
+    config: &'a CodeGeneratorConfig,
+    libraries: Vec<String>,
+}
+
+struct OCamlEmitter<'a, T> {
+    out: IndentedWriter<T>,
+    generator: &'a CodeGenerator<'a>,
+}
+
+impl<'a> CodeGenerator<'a> {
+    pub fn new(config: &'a CodeGeneratorConfig) -> Self {
+        if config.c_style_enums {
+            panic!("OCaml does not support generating c-style enums");
+        }
+        Self {
+            config,
+            libraries: config
+                .external_definitions
+                .keys()
+                .map(|k| k.to_string())
+                .collect::<Vec<_>>(),
+        }
+    }
+
+    pub fn output(&self, out: &mut dyn Write, registry: &Registry) -> Result<()> {
+        let mut emitter = OCamlEmitter {
+            out: IndentedWriter::new(out, IndentConfig::Space(2)),
+            generator: self,
+        };
+        emitter.output_preamble()?;
+        // todo: do a strongly connected components sort to regroup types that are cyclic
+        let n = registry.len();
+        for (i, (name, format)) in registry.iter().enumerate() {
+            let first = i == 0;
+            let last = i == n - 1;
+            emitter.output_container(name, format, first, last)?;
+        }
+        Ok(())
+    }
+}
+
+static KEYWORDS: phf::Set<&str> = phf_set! {
```

Comment:
Very nice. I didn't know `phf`.

Reply:
I found other ways also, but it looks like rewriting what's in the macro anyway

Reply:
I put `capitalize` in common.rs, removed the todo comment and added `(env (_ (flags (:standard -w -30-42))))` to the dune file